lm_eval/tasks/multiple.py

```diff
@@ -0,0 +1,176 @@
+import os
+import re
+import tempfile
+from multiprocessing import cpu_count
+from pathlib import Path
+from time import time
+
+import numpy as np
+from datasets import load_dataset
+from tqdm import tqdm
+
+from lm_eval.base import Task
+from lm_eval.tasks.custom_metrics.multiple_metrics.evaluation import \
+    evaluate_problem
+from lm_eval.tasks.custom_metrics.multiple_metrics.single_experiment_pass_k import \
+    for_file
+
+_CITATION = """
+@article{cassano2022scalable,
+  title={A Scalable and Extensible Approach to Benchmarking NL2Code for 18 Programming Languages},
+  author={Cassano, Federico and Gouwar, John and Nguyen, Daniel and Nguyen, Sydney and Phipps-Costin, Luna and Pinckney, Donald and Yee, Ming Ho and Zi, Yangtian and Anderson, Carolyn Jane and Feldman, Molly Q and others},
+  journal={arXiv preprint arXiv:2208.08227},
+  year={2022}
+}
+"""
+
+LANGUAGES = [
+    "py",
+    "bs",
+    "cpp",
+    "cs",
+    "d",
+    "go",
+    "java",
+    "js",
+    "jl",
+    "lua",
+    "pl",
+    "php",
+    "r",
+    "rkt",
+    "rb",
+    "rs",
+    "scala",
+    "swift",
+    "ts",
+]
+
+
+def create_all_tasks():
+    """Creates a dictionary of tasks from a list of levels
+    :return: {task_name: task}
+        e.g. {multiple-py: Task, multiple-java: Task}
+    """
+    return {f"multiple-{language}": create_task(language) for language in LANGUAGES}
+
+
+def create_task(language):
+    class MultiPLE(GeneralMultiPLE):
+        def __init__(self):
+            super().__init__(language)
+
+    return MultiPLE
+
+
+class GeneralMultiPLE(Task):
+    """A task represents an entire benchmark including its dataset, problems,
+    answers, generation settings and evaluation methods.
+    """
+
+    DATASET_PATH = "nuprl/MultiPL-E"
+    DATASET_NAME = None
+
+    def __init__(self, language):
+        self.language = language
+        self.DATASET_NAME = f"humaneval-{language}"
+        # we need the dataset to get stop words for each language
+        self.dataset = load_dataset(GeneralMultiPLE.DATASET_PATH, self.DATASET_NAME)
+        stop_words = self.dataset["test"][0]["stop_tokens"]
+        super().__init__(
+            stop_words=stop_words,
+            requires_execution=True,
+        )
+
+    def get_dataset(self):
+        """Returns dataset for the task or an iterable of any object, that get_prompt can handle"""
+        return self.dataset["test"]
+
+    def get_prompt(self, doc):
+        """Builds the prompt for the LM to generate from."""
+        return doc["prompt"].strip()
+
+    def get_reference(self, doc):
+        """Builds the reference solution for the doc (sample from the test dataset)."""
+        return doc["tests"]
+
+    @staticmethod
+    def remove_last_block(string, stop_words):
+        # Remove the last block of the code containing stop_words for HumanEval
+        string_list = re.split("(%s)" % "|".join(stop_words), string)
+        # last string should be ""
+        return "".join(string_list[:-2])
```

Comment:
I'm not quite sure what is happening here. I assume the `|stop_words` is added in some other file.

Reply:
yes `stop_words` are given to `model.generate` [here](https://github.com/loubnabnl/bigcode-evaluation-harness/blob/ba235616041eee8c056d62bbe2e77a90f7452db7/lm_eval/generation.py#L58)